articles/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/index.md

```diff
@@ -0,0 +1,849 @@
+
+- **chat**: The function we pass to `socket.on(“chat”)` handles sending and receiving message. Whenever a user disconnects, a message that the *"username has left the room"* will be broadcasted to all the people in the room.
+
+The event listeners of the above functions, `joinRoom` and `chat` will be triggered from the frontend in the files `home.js` and `chat.js` as explained later in this guide.
+
+### Creating the Frontend
+
+We will use React, [Redux library](https://redux.js.org/tutorials/essentials/part-1-overview-concepts), the socket.io-client, and aes256 for encrypting and decrypting the messages for the frontend.
+
+The folder structure for our client-side will appear as below:
+
+![frontend folder structure](/engineering-education/creating-a-simple-real-time-chat-app-with-react-socket-io-with-e2e-encryption/frontend-folders.PNG)
+
+### Coding the Frontend
+
+**Step 1**
+
+First, let's run the below commands in the terminal to create a client folder for our React App, namely `chatfrontend`, browse to the created directory and install the necessary dependencies required for react app to run.
+
+```bash
+npx create-react-app chatfrontend
+cd chatfrontend
+npm i node-sass react-redux react-router-dom redux socket.io-client aes256
+```
+
+**Step 2**
+
+Next, let's modify the file `/src/index.js` to help implement [reducers](https://www.geeksforgeeks.org/introduction-to-redux-action-reducers-and-store/) in our react app as explained later in this guide. The code will be as below:
+
+```typescript
+import App from "./App";
+import rootReducers from "./store/reducer/index";
+import ReactDOM from "react-dom";
+import { createStore } from "redux";
+import React from "react";
+import { Provider } from "react-redux";
+
+
+const store = createStore(rootReducers);
+ReactDOM.render(
+  <Provider store={store}>
+    <App />
+  </Provider>,
+  document.getElementById("root")
+);
+```
+
+In the above code, we have added `redux` and imported `reducers` from the file `/store/reducer/index.js`
+
+Next, let's create a file `/store/action/index.js` and add the code as shown below:
+
+```typescript
+export const process = (encrypt, text, cypher) => {
+  return {
+    type: "PROCESS",
+    payload: {
+      encrypt,
+      text,
+      cypher,
+    },
+  };
+};
+```
+
+Then let's create a file `/store/reducer/process.js` and add the code as shown below:
+
+```typescript
+export const ProcessReducer = (state = {}, action) => {
+  switch (action.type) {
+    case "PROCESS":
+      return { ...action.payload };
+
+    default:
+      return state;
+  }
+};
+```
+
+Then, let's create a file `/store/reducer/index.js` and add the code as shown below:
+
+```typescript
+import { ProcessReducer } from "./process";
+import { combineReducers } from "redux";
+const rootReducers = combineReducers({
+  ProcessReducer: ProcessReducer,
+});
+
+export default rootReducers;
+```
```

Comment:
Please provide a brief intro about these files and provide links to the concept of action and reducers.